{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '../')\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from random import seed\n",
    "import scvi\n",
    "from utils.utils import correction_validity, setup_logger, create_directory, nuclei_clustering, \\\n",
    "                        setup_reference_label, DEGs_HVGs, scIntegration, UMAP_visualization, save_anndata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# __Setup Non-Tuned Parameters (Unsupervised Model - [scVI](https://www.nature.com/articles/s41592-018-0229-2))__\n",
    "‚≠ê <u>SEED</u>: random seed, guarantee reproducibility of results;<br>\n",
    "‚≠ê <u>accelerator</u>: accelerator type - e.g., 'gpu';<br>\n",
    "\n",
    "‚≠ê <u>dir_adata</u>: directory of AnnData to be integrated, should be with dimension n_cell X n_genes [e.g., all protein coding genes];<br>\n",
    "‚≠ê <u>layer</u>: layers in AnnData, scVI and scANVI use raw counts;<br>\n",
    "‚≠ê <u>level</u>: cell type resolution level; e.g., level = 1 indicates all cell types, level = 2 indicates cell type subtypes, like Neuron subtypes, etc.;<br>\n",
    "‚≠ê <u>celltype</u>: celltype AnnData to be integrated;<br>\n",
    "‚≠ê <u>dataset_key</u>: column name in integrated AnnData that denotes dataset;<br>\n",
    "\n",
    "‚≠ê <u>dir_scvi</u>: directory of saving scVI (unsupervised) model;<br>\n",
    "‚≠ê <u>dir_hvg</u>: directory of saving highly variable genes;<br>\n",
    "‚≠ê <u>top_hvg_genes</u>: numbers of top highly variable genes considered;<br>\n",
    "‚≠ê <u>subset</u>: Inplace subset to highly-variable genes if True otherwise merely indicate highly variable genes;<br>\n",
    "‚≠ê <u>span</u>: The fraction of the data (cells) used when estimating the variance in the loess model fit if flavor='seurat_v3';<br>\n",
    "\n",
    "‚≠ê <u>use_unsupervised</u>: use unsupervised model (True -> scVI) or semi-supervised model (False -> scANVI);<br>\n",
    "\n",
    "‚≠ê <u>train_ratio</u>: ratio of training data, train_ratio + validation_test_ratio = 1;<br>\n",
    "‚≠ê <u>test_ratio</u>: ratio of test data considering of both validation and test data, e.g., train_ratio = 0.6, test_ratio = 0.5, implies training data is 60%, validation data is 20% and test data is 20%;<br>\n",
    "‚≠ê <u>shffule</u>: whether or not to shuffle the data before splitting;<br>\n",
    "‚≠ê <u>correction</u>: various types of corrections to be adjusted during integration, it can be 'no_batch_covariates': no batch and any covariates adjustments; 'batch_only': only adjust batch; 'covariates_only': only adjust covariates; 'batch_covariates': adjust both batch and covariates;<br>\n",
    "‚≠ê <u>batch_key</u>: batch column in AnnData.obs;<br>\n",
    "‚≠ê <u>categorical_covariate_keys</u>: list of categorical covariates to be adjusted during integration, e.g., sex, brain_region;<br>\n",
    "‚≠ê <u>continuous_covariate_keys</u>: list of continuous covariates to be adjusted during integration, e.g., QC factors.<br>\n",
    "\n",
    "‚≠ê <u>clustering_res</u>: resolution of cell/nuclei clustering.<br>\n",
    "‚≠ê <u>clustering_method</u>: clustering algorithm, could be leiden and louvain.<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEED = 1234\n",
    "accelerator = 'gpu'\n",
    "dir_adata = \"../test_Data/raw_count_before_integration.h5ad\"\n",
    "layer = 'counts'\n",
    "level = 1\n",
    "celltype = 'All'\n",
    "dataset_key = 'Dataset'\n",
    "dir_scvi = f'../output_test/Level{level}/{celltype}/scVI'\n",
    "dir_hvg = f'../output_test/Level{level}/{celltype}_HVG.txt'\n",
    "top_hvg_genes = 2000\n",
    "subset = True\n",
    "span = 1.0\n",
    "\n",
    "use_unsupervised = False\n",
    "\n",
    "train_ratio = 0.6\n",
    "test_ratio = 0.5\n",
    "shuffle = True\n",
    "correction = 'batch_covariates'\n",
    "batch_key = 'BatchID'\n",
    "categorical_covariate_keys = None\n",
    "continuous_covariate_keys = [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\", \"pct_counts_rb\"]\n",
    "\n",
    "\n",
    "correction_validity(correction = correction, \n",
    "                        batch_key = batch_key, \n",
    "                        continuous_covariate_keys = continuous_covariate_keys, \n",
    "                        categorical_covariate_keys = categorical_covariate_keys\n",
    "                    )\n",
    "\n",
    "clustering_resolution = 0.5\n",
    "clustering_method = 'leiden'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# __Setup Additional Non-Tuned Parameters (Semi-supervised Model - [scANVI](https://www.embopress.org/doi/full/10.15252/msb.20209620))__\n",
    "‚≠ê <u>dir_scanvi</u>: directory of saving scANVI (semi-supervised) model;<br>\n",
    "‚≠ê <u>labels_key</u>: label column in AnnData.obs, e.g., Column where you save the cell type labels;<br>\n",
    "‚≠ê <u>scanvi_epochs</u>: numbers of epochs for training scANVI (semi-supervised) model;<br>\n",
    "‚≠ê <u>dir_deg</u>: directory of saving differentially expressed genes (if using semi-supervised model);<br>\n",
    "‚≠ê <u>deg_group</u>: column in AnnData.obs that is used as the group variable when computing differentially expressed genes (if using semi-supervised model);<br>\n",
    "‚≠ê <u>top_deg_num</u>: numbers of top differentially expressed genes to be kept;<br>\n",
    "‚≠ê <u>use_sampling</u>: whether or not do downsampling when computing differentially expressed genes;<br>\n",
    "‚≠ê <u>sampling_ratio</u>: downsampling ratio when computing differentially expressed genes;<br>\n",
    "\n",
    "‚≠ê <u>prediction_key</u>: column in AnnData.obs that stores predicted cell type annotations;<br>\n",
    "‚≠ê <u>use_reference</u>: whether or not use reference cell labels for integration;<br>\n",
    "‚≠ê <u>reference_datasets</u>: list of reference datasets to be used, reference_datasets is more preferred when you have multiple reference datasets and have not been merged;<br>\n",
    "‚≠ê <u>dir_reference</u>: directory of reference datasets, dir_reference is more preferred when you already have a reference_adata.h5ad for example;<br>\n",
    "‚≠ê <u>ref_label</u>: column in AnnData.obs that indicates the reference cell type labels;<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dir_scanvi = f'../output_test/Level{level}/{celltype}/scANVI' if not use_unsupervised else None\n",
    "labels_key = 'seed_label' if not use_unsupervised else None\n",
    "scanvi_epochs = 5 if not use_unsupervised else None\n",
    "dir_deg = f'../output_test/Level{level}/{celltype}_DEG.pkl' if not use_unsupervised else None\n",
    "deg_group = labels_key if not use_unsupervised else None\n",
    "top_deg_num = 500 if not use_unsupervised else None\n",
    "use_sampling = True if not use_unsupervised else None\n",
    "sampling_ratio = 0.1 if not use_unsupervised else None\n",
    "if use_sampling:\n",
    "    assert sampling_ratio > 0 and sampling_ratio < 1, 'Choose the correct sampling ratio.'\n",
    "\n",
    "prediction_key = 'C_scANVI' if not use_unsupervised else None\n",
    "use_reference = True if not use_unsupervised else None\n",
    "reference_datasets = ['37824655', '37824663']\n",
    "dir_reference = None\n",
    "ref_label = 'Super_Celltype' if not use_unsupervised else None\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# __Setup Tuned HyperParameters__\n",
    " \n",
    "‚≠ê <u>enable_parameter_tuning</u>: boolean variable, indicating whether to tune hyperparameters or not. If               enable_parameter_tuning = False, hyperparameters are set to default values;<br>\n",
    "‚≠ê <u>num_random_trials</u>: numbers of random trials when performing hyperparameter tuning;<br>\n",
    "‚≠ê <u>n_latent</u>: dimension of latent embeddings;<br>\n",
    "‚≠ê <u>n_hidden</u>: dimensions of hidden layers;<br>\n",
    "‚≠ê <u>n_layers</u>: numbers of hidden layers;<br>\n",
    "‚≠ê <u>dispersion</u>: dispersion pattern of gene expression;<br>\n",
    "‚≠ê <u>lr</u>: learning rate;<br>\n",
    "‚≠ê <u>max_epochs</u>: numbers of epochs to be used to train scVI (unsupervised) model.<br>\n",
    "\n",
    "### __All parameters to be tuned follow the same <u>Dictionary</u> data structure.__ \n",
    "\n",
    "üîπ <u>default</u>: parameter default value;<br>\n",
    "üîπ <u>continuous low</u>: setup it when parameter is assumed to be continuous, and <u>continuous_low</u> denotes the lower bound during tuning;<br>\n",
    "üîπ <u>continuous high</u>: setup it when parameter is assumed to be continuous, and <u>continuous_high</u> denotes the upper bound during tuning;<br>\n",
    "üîπ <u>categorical_choices</u>: setup it when parameter is assumed to be categorical, put all categorical values in a list;<br>\n",
    "üîπ <u>log</u>: log = True when parameter is assumed to be continuous, otherwise log = False.<br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enable_parameter_tuning = False\n",
    "nums_random_trials = 5\n",
    "\n",
    "n_latent = {\n",
    "    \"default\": 50,\n",
    "    \"continuous_low\": None,\n",
    "    \"continuous_high\": None,\n",
    "    \"categorical_choices\": [10, 30, 50],\n",
    "    \"log\": False\n",
    "}\n",
    "\n",
    "n_hidden = {\n",
    "    \"default\": 64,\n",
    "    \"continuous_low\": None,\n",
    "    \"continuous_high\": None,\n",
    "    \"categorical_choices\": [64, 128, 256],\n",
    "    \"log\": False\n",
    "}\n",
    "\n",
    "n_layers = {\n",
    "    \"default\": 1,\n",
    "    \"continuous_low\": None,\n",
    "    \"continuous_high\": None,\n",
    "    \"categorical_choices\": [1, 2, 3],\n",
    "    \"log\": False\n",
    "}\n",
    "\n",
    "\n",
    "dispersion = {\n",
    "    \"default\": \"gene\",\n",
    "    \"continuous_low\": None,\n",
    "    \"continuous_high\": None,\n",
    "    \"categorical_choices\": [\"gene\", \"gene-batch\"],\n",
    "    \"log\": False\n",
    "}\n",
    "\n",
    "\n",
    "lr = {\n",
    "    \"default\": 5e-4,\n",
    "    \"continuous_low\": 1e-5,\n",
    "    \"continuous_high\": 1e-3,\n",
    "    \"categorical_choices\": None,\n",
    "    \"log\": True\n",
    "}\n",
    "\n",
    "max_epochs = {\n",
    "    \"default\": 10,\n",
    "    \"continuous_low\": None,\n",
    "    \"continuous_high\": None,\n",
    "    \"categorical_choices\": [50, 100, 200],\n",
    "    \"log\": False\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# seed everything\n",
    "seed(SEED)\n",
    "scvi.settings.seed = SEED\n",
    "\n",
    "logger = setup_logger(name = \"TACA\", \n",
    "                    log_file = f'{celltype}.log',\n",
    "                    log_dir = f'../output_test/Level{level}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load and preprocess AnnData to be integarated\n",
    "\n",
    "adata = ad.read_h5ad(dir_adata)\n",
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "\n",
    "logger.info(\"AnnData has {:,} cell/nuclei and {:,} genes.\".format(adata.shape[0], adata.shape[1]))\n",
    "\n",
    "if not use_unsupervised:\n",
    "    adata = setup_reference_label(\n",
    "        adata = adata,\n",
    "        labels_key = labels_key,\n",
    "        ref_label = ref_label,\n",
    "        celltype = celltype,\n",
    "        dataset_key = dataset_key,\n",
    "        reference_datasets = reference_datasets,\n",
    "        dir_reference = dir_reference\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create directory to save unsupervised (scVI) and semi-supervised (scANVI) models\n",
    "\n",
    "create_directory(dir_scvi)\n",
    "\n",
    "if dir_scanvi is not None:\n",
    "    create_directory(dir_scanvi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function DEGs_HVGs\n",
    "\n",
    "- Compute highly variable genes (HVGs) and differential expressed genes (DEGs, if using semi-supervised model);\n",
    "\n",
    "- Subset anndata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "adata = DEGs_HVGs(\n",
    "        adata = adata, \n",
    "        dir_deg = dir_deg, \n",
    "        dir_hvg = dir_hvg,\n",
    "        labels_key = labels_key,\n",
    "        deg_group = deg_group, \n",
    "        top_deg_num = top_deg_num,\n",
    "        use_sampling = use_sampling,\n",
    "        sampling_ratio = sampling_ratio,\n",
    "        use_reference = use_reference,\n",
    "        top_hvg_genes = top_hvg_genes,\n",
    "        subset = subset,\n",
    "        layer = layer,\n",
    "        batch_key = batch_key,\n",
    "        span = span\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function scIntegration\n",
    "\n",
    "- Integration with either unsupervised ([scVI](https://www.nature.com/articles/s41592-018-0229-2)) or semi-supervised ([scANVI](https://www.embopress.org/doi/full/10.15252/msb.20209620)) model\n",
    "\n",
    "    1. split adata into adata_train, adata_val, adata_test using train_ratio and test_ratio predefined;\n",
    "\n",
    "    2. tune hyperparameter parameters based on evidence lower bound (ELBO) value, and generate best hyperparameters;\n",
    "\n",
    "    3. generate entire nuclei embedding (and cell type predictions if using semi-supervised model) using best hyperparameters.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = scIntegration(\n",
    "    adata = adata,\n",
    "    level = level,\n",
    "    train_ratio = train_ratio,\n",
    "    test_ratio = test_ratio,\n",
    "    seed_value = SEED,\n",
    "    shuffle = shuffle,\n",
    "    adjust_cols = [batch_key],\n",
    "    accelerator = accelerator,\n",
    "    labels_key = labels_key,\n",
    "    prediction_key = prediction_key,\n",
    "    layer = layer,\n",
    "    batch_key = batch_key,\n",
    "    categorical_covariate_keys = categorical_covariate_keys,\n",
    "    continuous_covariate_keys = continuous_covariate_keys,\n",
    "    enable_parameter_tuning = enable_parameter_tuning,\n",
    "    n_latent = n_latent if enable_parameter_tuning else n_latent[\"default\"],\n",
    "    n_hidden = n_hidden if enable_parameter_tuning else n_hidden[\"default\"],\n",
    "    n_layers = n_layers if enable_parameter_tuning else n_layers[\"default\"],\n",
    "    dispersion = dispersion if enable_parameter_tuning else dispersion[\"default\"],\n",
    "    lr = lr if enable_parameter_tuning else lr[\"default\"],\n",
    "    max_epochs = max_epochs if enable_parameter_tuning else max_epochs[\"default\"],\n",
    "    scanvi_epochs = scanvi_epochs,\n",
    "    dir_scvi = dir_scvi,\n",
    "    dir_scanvi = dir_scanvi,\n",
    "    use_unsupervised = use_unsupervised,\n",
    "    nums_random_trials = nums_random_trials\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## __Function nuclei_clustering__\n",
    "- Cluster all nuclei with learned embedding using either 'leiden' or 'louvain' clustering algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = nuclei_clustering(\n",
    "    adata = adata,\n",
    "    level = level,\n",
    "    method = clustering_method,\n",
    "    resolution = clustering_resolution,\n",
    "    seed_value = SEED,\n",
    "    use_unsupervised = use_unsupervised\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## __Function UMAP_visulization__\n",
    "- Generate 2D UMAP embedding for visulization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = UMAP_visualization(\n",
    "    adata = adata,\n",
    "    seed_value = SEED,\n",
    "    level = level,\n",
    "    use_unsupervised = use_unsupervised\n",
    ")\n",
    "\n",
    "plt.figure(figsize = (6,6))\n",
    "sc.pl.umap(adata)\n",
    "plt.savefig('./test1.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## __Function save_anndata__\n",
    "- Load AnnData object before integration (dir_adata), should be with dimension n_cell X n_genes [e.g., all protein coding genes];<br>\n",
    "- Add cell type prediction if any;\n",
    "- Add cell embeddings from unsupervised model (e.g., scVI);\n",
    "- Add umap embeddings;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_anndata(\n",
    "    dir_adata = dir_adata, \n",
    "    adata = adata,\n",
    "    level = level,\n",
    "    prediction_key = prediction_key,\n",
    "    use_unsupervised = use_unsupervised,\n",
    "    dir_save = dir_scvi if use_unsupervised else dir_scanvi,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
